parser/prism/Helpers.cc

```diff
@@ -0,0 +1,783 @@
+    call->block = up_cast(block);
+
+    return send;
+}
+
+pm_node_t *PMK::Array(core::LocOffsets loc, const vector<pm_node_t *> &elements) {
+    // Create an array node
+    pm_array_node_t *array = allocateNode<pm_array_node_t>();
+    if (!array) {
+        return nullptr;
+    }
+
+    pm_node_t **elemNodes = nullptr;
+    size_t elementsSize = elements.size();
+    if (!elements.empty()) {
+        elemNodes = copyNodesToArray(elements);
+        if (!elemNodes) {
+            free(array);
+            return nullptr;
+        }
+    }
+
+    *array = (pm_array_node_t){.base = initializeBaseNode(PM_ARRAY_NODE),
+                               .elements = {.size = elementsSize, .capacity = elementsSize, .nodes = elemNodes},
+                               .opening_loc = convertLocOffsets(loc.copyWithZeroLength()),
+                               .closing_loc = convertLocOffsets(loc.copyEndWithZeroLength())};
+
+    array->base.location = convertLocOffsets(loc);
+
+    return up_cast(array);
+}
+
+pm_node_t *PMK::T_Array(core::LocOffsets loc) {
+    // Create T::Array constant path
+    return ConstantPathNode(loc, T(loc), "Array");
+}
+
+pm_node_t *PMK::T_Class(core::LocOffsets loc) {
+    // Create T::Class constant path
+    return ConstantPathNode(loc, T(loc), "Class");
+}
+
+pm_node_t *PMK::T_Enumerable(core::LocOffsets loc) {
+    // Create T::Enumerable constant path
+    return ConstantPathNode(loc, T(loc), "Enumerable");
+}
+
+pm_node_t *PMK::T_Enumerator(core::LocOffsets loc) {
+    // Create T::Enumerator constant path
+    return ConstantPathNode(loc, T(loc), "Enumerator");
+}
+
+pm_node_t *PMK::T_Hash(core::LocOffsets loc) {
+    // Create T::Hash constant path
+    return ConstantPathNode(loc, T(loc), "Hash");
+}
+
+pm_node_t *PMK::T_Set(core::LocOffsets loc) {
+    // Create T::Set constant path
+    return ConstantPathNode(loc, T(loc), "Set");
+}
+
+pm_node_t *PMK::T_Range(core::LocOffsets loc) {
+    // Create T::Range constant path
+    return ConstantPathNode(loc, T(loc), "Range");
+}
+
+bool PMK::isTUntyped(pm_node_t *node) {
+    if (!node || PM_NODE_TYPE(node) != PM_CALL_NODE) {
+        return false;
+    }
+
+    pm_call_node_t *call = down_cast<pm_call_node_t>(node);
+    if (!call->receiver || PM_NODE_TYPE(call->receiver) != PM_CONSTANT_PATH_NODE) {
+        return false;
+    }
+
+    // Check if receiver is ::T and method is "untyped"
+    pm_constant_path_node_t *receiver = down_cast<pm_constant_path_node_t>(call->receiver);
+    if (receiver->parent != nullptr) {
+        return false; // Should be root-anchored ::T
+    }
+
+    auto methodName = parser()->resolveConstant(call->name);
+    if (methodName != "untyped") {
+        return false;
+    }
+
+    auto receiverName = parser()->resolveConstant(receiver->name);
+    return receiverName == "T";
+}
+
+bool PMK::isSetterCall(pm_node_t *node, const Parser &parser) {
+    if (PM_NODE_TYPE(node) != PM_CALL_NODE) {
```

Comment:
Please check for any use of `PM_NODE_TYPE\(.*\) [!=]=` and replace with `PM_NODE_TYPE_P`:

```suggestion
    if (!PM_NODE_TYPE(node, PM_CALL_NODE)) {
```